pkg/egresspolicy/manager.go

```diff
@@ -0,0 +1,461 @@
+	defer manager.mutex.Unlock()
+	// Ns labels have been updated, need to re-evaluate all pods within ns.
+	pods := manager.getLocalPodsForNamespace(ns)
+	if len(pods) == 0 {
+		return
+	}
+	for _, pod := range pods {
+		manager.OnUpdatePodLocked(pod, true /* removeOld */, true /* upsertNew */)
+	}
+}
+
+func (manager *Manager) getPodMetadata(pod *slim_corev1.Pod, podIPs []string) (*podMetadata, error) {
+	podId := types.NamespacedName{
+		Name:      pod.GetName(),
+		Namespace: pod.GetNamespace(),
+	}
+
+	namespace, err := manager.getNamespaceForPod(pod)
+	if namespace == nil || err != nil {
+		data := &podMetadata{
+			ips:    podIPs,
+			labels: pod.GetLabels(),
+			id:     podId,
+		}
+
+		// Keep id to pod data mapping up to date
+		manager.podDataStore[podId] = data
+		return data, nil
+	}
+
+	data := &podMetadata{
+		ips:      podIPs,
+		labels:   pod.GetLabels(),
+		nsLabels: namespace.GetLabels(),
+		id:       podId,
+	}
+
+	// Keep id to pod data mapping up to date
+	manager.podDataStore[podId] = data
+	return data, nil
+}
+
+func (manager *Manager) getNamespaceForPod(pod *slim_corev1.Pod) (*slim_corev1.Namespace, error) {
+	nsName := pod.GetNamespace()
+	nsStore := manager.storeGetter.GetStore("namespace")
+	nsItem, exists, _ := nsStore.GetByKey(pod.GetNamespace())
+	if !exists {
+		log.WithFields(logrus.Fields{logfields.K8sNamespace: nsName}).Warn("Namespace does not exist.")
+		return nil, nil
+	}
+	return nsItem.(*slim_corev1.Namespace), nil
+}
+
+// isValidConfig validates the given policy config.
+func (manager *Manager) isValidConfig(config EgressPolicyConfig) error {
+	for _, policyConfig := range manager.policyConfigs {
+		if policyConfig.egressIP.String() == config.egressIP.String() {
+			return fmt.Errorf("CiliumEgressNatPolicy for egress IP %v already exists", config.egressIP.String())
+		}
+	}
+	return nil
+}
+
+// upsertPolicyPod updates or inserts to pod policy mapping for given policy config and pods,
+// it also upserts the pod data cache and egress map to keep in sync
+func (manager *Manager) upsertPolicyPod(config *EgressPolicyConfig, pods ...*podMetadata) error {
+	for _, pod := range pods {
+		podIp := net.ParseIP(pod.ips[0])
+		if podPolicies, ok := manager.policyPods[pod.id]; ok {
+			newPodPolicy := true
+			for _, poID := range podPolicies {
+				// Existing pod policy update, noop for this policy
+				if poID == config.id {
+					newPodPolicy = false
+					break
+				}
+			}
+			if newPodPolicy {
+				// Pod selected by a new policy
+				err := manager.updateEgressMap(podIp, config)
+				if err != nil {
+					return err
+				}
+				manager.policyPods[pod.id] = append(manager.policyPods[pod.id], config.id)
+				manager.podDataStore[pod.id] = pod
+			}
+		} else {
+			// Pod selected by a policy for the first time
+			err := manager.updateEgressMap(podIp, config)
+			if err != nil {
+				return err
+			}
+			pp := []policyID{config.id}
+			manager.policyPods[pod.id] = pp
+			manager.podDataStore[pod.id] = pod
+		}
```

Comment:
I don't understand why this entire code isn't simply:
```golang
			err := manager.updateEgressMap(podIp, config)
			if err != nil {
				return err
			}
```

there's no need to keep the state if a pod has or hasn't already a policy in place.

Reply:
Well other than updating the map, we also need to update the endpoint policy mapping. Otherwise when we delete an endpoint or a policy it's hard to know what to delete in the map.
Let me know if there is something else i missed here.

Reply:
I refactored this part of code. It should be a lot clearer now. Let me know if it looks good to ya.
Thx!